ToDo.Core/TaskController.cpp

```diff
@@ -0,0 +1,72 @@
+#include "pch.h"
+#include <future>
+
+#include "TaskController.h"
+#include "ToDoTask.h"
+
+namespace ToDo::Core {
+API::IToDoTask &CreateTask(long id, std::string description,
+                           bool hasCompleted) {
+  return *(new ToDoTask(id, description, hasCompleted));
+}
+
+TaskController::TaskController(API::ITaskStorage &taskStorage) {
+  _taskStorage = &taskStorage;
+}
+
+std::optional<std::reference_wrapper<API::IToDoTask>>
+TaskController::AddTask(std::string description) {
+  if (description.empty()) {
+    return std::nullopt;
+  }
+
+  auto getNewAddedTask =
+      std::async(std::launch::async, [&]() -> API::IToDoTask & {
+        if (!_newIdForUse) {
+          _newIdForUse = _taskStorage->GetTotalNumberOfTasks();
+        }
+
+        return _taskStorage->AddNewTask((*_newIdForUse)++, description,
+                                        CreateTask);
+      });
+
+  return std::optional<std::reference_wrapper<API::IToDoTask>>{
+      getNewAddedTask.get()};
+}
+
+void TaskController::OnTaskAttributesChanged(ToDoTask &sender,
+                                             TaskAttribute changedAttribute) {
+  switch (changedAttribute) {
+  case TaskAttribute::Description:
+    _taskStorage->UpdateTaskDescription(sender.GetId(),
+                                        sender.GetDescription());
+    break;
+
+  case TaskAttribute::HasCompleted:
+    _taskStorage->UpdateTaskHasCompleted(sender.GetId(),
+                                         sender.GetHasCompleted());
+    break;
+  }
+}
+
+std::vector<std::reference_wrapper<API::IToDoTask>>
+TaskController::GetAlltasks() {
+  auto getAllTasks = std::async(
+      std::launch::async,
+      [&]() -> std::vector<std::reference_wrapper<API::IToDoTask>> {
```

Comment:
In newer C++ the return-type for a lambda is optional.

Reply:
https://github.com/Take-A-Byte/ToDoApp/pull/1/commits/0df14db7965f5b544fae7c424c67d15feb3e48ab